experimental/personnames/src/derive_locale.rs

```diff
@@ -0,0 +1,228 @@
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use icu_locid::{subtags, Locale};
+use icu_properties::script::ScriptWithExtensionsBorrowed;
+use subtags::Script;
+
+use crate::api::NameFieldKind::{Given, Surname};
+use crate::api::{NameFieldKind, PersonName, PersonNamesFormatterError};
+
+/// Override the formatting payload to use based on specification rules.
+///
+/// if name locale and formatting locale are incompatible, name locale takes precedence
+/// it should dynamically load the name locale formatter using the data_provider given in constructor.
+/// https://www.unicode.org/reports/tr35/tr35-personNames.html#switch-the-formatting-locale-if-necessary
+///
+/// The formatter locale and name locale must be maximized first.
+///
+pub(crate) fn effective_locale<'a>(
+    formatter_locale: &'a Locale,
+    person_name_locale: &'a Locale,
+) -> Result<&'a Locale, PersonNamesFormatterError> {
+    let name_script = person_name_locale.id.script.unwrap();
+    let formatter_script = formatter_locale.id.script.unwrap();
+    if !compatible_scripts(name_script, formatter_script) {
+        return Ok(person_name_locale);
+    }
+    Ok(formatter_locale)
+}
+
+// TODO: proper handling of compatible scripts.
+fn compatible_scripts(sc1: Script, sc2: Script) -> bool {
+    let sc1_str = sc1.as_str();
+    let sc2_str = sc2.as_str();
+    let jpan_compatible = ["Hani", "Kana", "Hira"];
+    if sc1_str == "Jpan" && jpan_compatible.contains(&sc2_str) {
+        return true;
+    }
+    if sc2_str == "Jpan" && jpan_compatible.contains(&sc1_str) {
+        return true;
+    }
+    sc1_str == sc2_str
+}
+
+/// https://www.unicode.org/reports/tr35/tr35-personNames.html#derive-the-name-locale
+pub(crate) fn likely_person_name_locale<N>(
+    person_name: &N,
+) -> Result<Locale, PersonNamesFormatterError>
+where
+    N: PersonName,
+{
+    use icu_properties::Script;
+
+    let swe = icu_properties::script::script_with_extensions();
+    let mut found_name_script = find_script(person_name, swe, Surname);
+    if found_name_script.is_none() {
+        found_name_script = find_script(person_name, swe, Given);
+    }
+    let name_script = found_name_script.unwrap_or(Script::Unknown);
+
+    let lookup = Script::enum_to_short_name_mapper().static_to_owned();
+    let lookup = lookup.as_borrowed();
+
+    person_name.name_locale().map_or_else(
+        || {
+            let default_language_str =
+                "und-".to_owned() + lookup.get(name_script).unwrap().as_str();
+            let effective_locale: Locale = default_language_str.parse().map_err(|_err| {
+                PersonNamesFormatterError::ParseError(String::from(
+                    "Cannot build default locale for parser",
+                ))
+            })?;
+            Ok(effective_locale)
+        },
+        |locale| {
+            let mut effective_locale = locale.clone();
+            let effective_script = lookup
+                .get(name_script)
```

Comment:
At this point you have a `TinyAsciiStr<4>`. You should be able to convert it a bit more efficiently by using Script::[try_from_raw](https://unicode-org.github.io/icu4x/docs/icu/locid/subtags/struct.Script.html#method.try_from_raw)